scripts/shtf.py

```diff
@@ -0,0 +1,629 @@
+
+
+class MetaType(TypedDict):
+    request_id: str
+    message_type: str
+    message_link: str
+
+
+class NotificationType(TypedDict):
+    meta: MetaType
+    body: str
+
+
+@dataclass
+class Stats:
+    fills: int = 0
+    requests: int = 0
+    withdraws: int = 0
+    volume: int = 0
+
+
+@dataclass
+class TransferStats:
+    volume: int = 0
+    fills: int = 0
+    requests: int = 0
+    withdrawals: int = 0
+    claims: int = 0
+    expired_requests: int = 0
+    expired_requests_volume: int = 0
+
+
+class NotificationState:
+    def __init__(self) -> None:
+        with open(os.path.join(os.path.dirname(__file__), "breadcrumbs.json"), "r") as f:
+            self.notified_state = json.load(f)
+
+    def is_set(self, request_id: str, notified_on: str) -> bool:
+        breadcrumb = self.notified_state.get(str(request_id))
+        if breadcrumb is None:
+            return False
+        else:
+            return breadcrumb.get(notified_on, False)
+
+    def update(self, request_id: str, notified_on: str) -> None:
+        breadcrumb = self.notified_state.get(str(request_id))
+        if breadcrumb is None:
+            self.notified_state[str(request_id)] = dict()
+            self.notified_state[str(request_id)][notified_on] = True
+        else:
+            breadcrumb[notified_on] = True
+
+    def persist(self) -> None:
+        with open(os.path.join(os.path.dirname(__file__), "breadcrumbs.json"), "w") as f:
+            json.dump(self.notified_state, f)
+
+
+@dataclass
+class Context:
+    stats: TransferStats = field(default_factory=TransferStats)
+    messages: list = field(default_factory=list)
+    agent_address: str = ""
+
+    notification_state: NotificationState = field(default_factory=NotificationState)
+
+    def add_message(self, message: NotificationType) -> None:
+        self.messages.append(message)
+
+
+class Notify:
+    def __init__(self, system: str, config: dict[str, dict[str, str]]) -> None:
+        self.system = system
+        self.config = config
+
+    def send(self, message: MessageType, callback: Optional[Callable] = None) -> None:
+        global notifications_sent
+        if notifications_sent > 0 and notifications_sent % 5 == 0:
+            time.sleep(timeout_between_sends)
+
+        text = message["text"]
+
+        if message.get("message_link"):
+            text += "\n" + f"View on [explorer]({message['message_link']})"
+
+        if self.system == "rocketchat":
+            status = self.send_to_rocketchat(text)
+        elif self.system == "telegram":
+            status = self.send_to_telegram(text)
+        else:
+            raise ValueError("Unknown notification system")
+
+        if status:
+            notifications_sent += 1
+            if callback is not None:
+                callback()
+
+    def send_to_rocketchat(self, text: str) -> bool:
+        headers = {"Content-Type": "application/json"}
+
+        body = {
+            "username": "beamer-monitoring-bot",
```

Comment:
is this just sending it? i cant see any authentication

Reply:
You don't need authentication. This is a webhook url - anyone could send to it, if you knew the url.